### PR TITLE
protect watch with auth

### PR DESCRIPTION
**Pull request by mitake**
Fix https://github.com/coreos/etcd/issues/7789

@heyitsanthony how do you think about this change? I couldn't separate auth and watch in v3rpc in a clean manner (like maintenance RPCs) because its stream RPC structure was hard to decompose. I'd like to hear your opinion.

### CODE REVIEW

File: etcdserver/api/v3rpc/watch.go [hunk 1]
```diff
@@ -180,6 +200,19 @@ func (sws *serverWatchStream) recvLoop() error {
 				// support  >= key queries
 				creq.RangeEnd = []byte{}
 			}
+
+			if !sws.isWatchPermitted(creq) {
+				wr := &pb.WatchResponse{
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: This should fill out all of the fields like the regular error path below. Older clients will be confused by this behavior since they'll get id=0 and not know about permission_denied and think the watch was successfully created.

File: clientv3/watch.go [hunk 2]
```diff
@@ -462,6 +471,9 @@ func (w *watchGrpcStream) run() {
 					close(ws.recvc)
 					closing[ws] = struct{}{}
 				}
+			case pbresp.PermissionDenied:
```
Comment (heyitsanthony) [inline clientv3/watch.go]: this should be handled through the `pbresp.Created` path if it's only going to happen on create events; it should work as-is if returning a createresponse with canceled=true and id=-1.

File: etcdserver/etcdserverpb/rpc.proto [hunk 3]
```diff
@@ -653,6 +653,9 @@ message WatchResponse {
   // The client should treat the watcher as canceled and should not try to create any
   // watcher with the same start_revision again.
   int64 compact_revision  = 5;
+  // permissionDenied is set to true if the requesting user doesn't have a
+  // read permission of watched keys.
+  bool permission_denied = 6;
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: I'm hesitant to include boolean flags to indicate errors since errors will be mutually exclusive. Maybe a string that can be translated back to a grpc error?

File: clientv3/client.go [hunk 4]
```diff
@@ -93,7 +93,10 @@ func NewFromURL(url string) (*Client, error) {
 // Close shuts down the client's etcd connections.
 func (c *Client) Close() error {
 	c.cancel()
-	c.Watcher.Close()
+	err := c.Watcher.Close()
+	if err != nil {
```
Comment (heyitsanthony) [inline clientv3/client.go]: `Close` should probably keep going regardless of whether if there's an error, otherwise it'll leak lease resources if there's any watch error

File: clientv3/watch.go [hunk 5]
```diff
@@ -324,6 +329,10 @@ func (w *watchGrpcStream) Close() (err error) {
 	case err = <-w.errc:
 	default:
 	}
+
+	if w.closeErr != nil {
```
Comment (heyitsanthony) [inline clientv3/watch.go]: why is this needed now?

File: e2e/ctl_v3_auth_test.go [hunk 6]
```diff
@@ -661,3 +662,80 @@ func authTestFromKeyPerm(cx ctlCtx) {
...
+	cx.user, cx.pass = "test-user", "pass"
+	for i, tt := range tests {
+		donec := make(chan struct{})
+		go func(i int, puts []kv) {
+			for j := range puts {
+				if err := ctlV3Put(cx, puts[j].key, puts[j].val, ""); err != nil {
+					cx.t.Fatalf("watchTest #%d-%d: ctlV3Put error (%v)", i, j, err)
+				}
+			}
+			close(donec)
```
Comment (heyitsanthony) [inline e2e/ctl_v3_auth_test.go]: `defer close(donec)` at top of goroutine; otherwise the fatal will cause the test to wait until `go test` timeout

File: e2e/ctl_v3_watch_test.go [hunk 7]
```diff
@@ -116,3 +116,30 @@ func ctlV3Watch(cx ctlCtx, args []string, kvs ...kv) error {
 	}
 	return proc.Stop()
 }
+
```
Comment (heyitsanthony) [inline e2e/ctl_v3_watch_test.go]: share setup code from `cmdArgs :=` to `if cx.interactive { ... }` instead of duplicating: `ctlV3WatchProc(cx ctlCtx, args []string) (*expect.ExpectProcess, error)`

File: etcdserver/etcdserverpb/rpc.proto (hunk 3 above)
Comment (mitake) [inline etcdserver/etcdserverpb/rpc.proto]: Thanks for your review and sorry for my late reply. I added the field (and other changes you pointed out problems) for expressing the error of permission denied. But as you say, if it can be expressed in the error code, it would be more straightfoward. I'll change the error propagation style in the next update.

**Comment by xiang90** [comment]
@mitake kindly ping. Can we get this in soon? it blocks our 3.2 release. The approach looks good.

**Comment by mitake** [comment]
@xiang90 sorry for that, I'll update this PR today!

**Comment by mitake** [comment]
@heyitsanthony @xiang90 In this version, I tried to propagate the permission denied error via `WatchResponse`. As @heyitsanthony pointed out, it is overkill and the error should be propagated an ordinary error.

I think `watchGrpcStream.closeErr` would be a suitable way for the propagation. But they are removed too early here: https://github.com/coreos/etcd/blob/master/clientv3/watch.go#L412 so the path of `watcher.Close()` cannot obtain the error with the variable. Is it ok to defer deleting the streams? I think `w.streams` shouldn't be deleted here https://github.com/coreos/etcd/blob/master/clientv3/watch.go#L334-L336 because `watcher.Close()` should collect errors from the streams here: https://github.com/coreos/etcd/blob/master/clientv3/watch.go#L312-L316

I'm still thinking about this change breaks semantics of clientv3. I'd like to hear your comments.

**Comment by mitake** [comment]
https://github.com/mitake/etcd/commit/e1103ccbd6682c87383b5d5c8ce691ed3d0e9807
This is a WIP version of my idea.

**Comment by mitake** [comment]
@heyitsanthony @xiang90 Anyway, I updated this PR based on your feedback. If the way of error propagation isn't suitable, I'll revisit it again.

File: clientv3/client.go [hunk 8]
```diff
@@ -93,8 +93,11 @@ func NewFromURL(url string) (*Client, error) {
 // Close shuts down the client's etcd connections.
 func (c *Client) Close() error {
 	c.cancel()
-	c.Watcher.Close()
+	closeErr := c.Watcher.Close()
```
Comment (mitake) [inline clientv3/client.go]: This line is breaking `TestGRPCResolver` and `TestGRPCResolverMulti` because it changes behaviour of `integration.ClusterV3.Terminate()`...

Anyway, current `Client.Close()` ignores an error from `c.Watcher.Close()`, do you have good idea for receiving the watcher error? @heyitsanthony @xiang90

**Comment by heyitsanthony** [comment]
@mitake an auth error probably shouldn't take down the entire watch stream (e.g., by returning an error on the stream); there would have to be an extra error path outside of watch channels just for handling auth errors. Sending a response with  `Canceled = true` from the etcd server on watch create should be enough to close down the watch (although the reason for the cancel won't be clear)

**Comment by mitake** [comment]
@heyitsanthony I see, then as you mentioned in the comment (https://github.com/coreos/etcd/pull/7809#discussion_r113293388), a new member for indicating the reason should be in `WatchResponse`. I think string is too generic, how about adding a new enum?

**Comment by mitake** [comment]
@heyitsanthony I updated with the enum approach, how about this?

**Comment by mitake** [comment]
@heyitsanthony sorry it seems that I misunderstood your intention. An auth error shouldn't do closing an entire watch stream like this: https://github.com/coreos/etcd/pull/7809/files#diff-bdef56e462f289cae8e13ebbe16ba455R476 ? The error should be handled in the receiver side, is this ok?

**Comment by mitake** [comment]
@heyitsanthony @xiang90 I let clientv3 return canceled events to its client. I think the latest PR can handle the permission denied error in a suitable manner and preserve the semantics of existing watch APIs. Could you take a look?

**Comment by heyitsanthony** [comment]
@mitake using `string` for the error is important here since the strings can map directly to errors etcd already has via `rpctypes.Error()`. Adding an enum for watch errors means another path for translating rpc errors into client-side error types; it's simpler to return arbitrary etcd errors because it avoids the need to handle a separate class of errors...

File: clientv3/watch.go [hunk 9]
```diff
@@ -456,11 +462,23 @@ func (w *watchGrpcStream) run() {
 					wc.Send(ws.initReq.toPB())
 				}
 			case pbresp.Canceled:
-				delete(cancelSet, pbresp.WatchId)
-				if ws, ok := w.substreams[pbresp.WatchId]; ok {
-					// signal to stream goroutine to update closingc
-					close(ws.recvc)
-					closing[ws] = struct{}{}
+				if pbresp.CancelReason == pb.WatchResponse_PERMISSION_DENIED {
```
Comment (heyitsanthony) [inline clientv3/watch.go]: does this really need special handling? why isn't the regular cancel=true path enough?

Comment (mitake) [inline clientv3/watch.go]: I thought the special handling is required because the error of permission denied will be replied from etcd to the client before registering a substream so the new `dispatchCanceledEvent()` is needed. Otherwise, the client cannot know the reason of the cancel. In a case of etcdctl, the command cannot print a detailed reason for a user.

**Comment by mitake** [comment]
@heyitsanthony ok then I'll change the type of the filed for cancel reason from enum to string in the next update.

**Comment by mitake** [comment]
@heyitsanthony updated for changing the reason field from enum to string. Could you take a look?

File: clientv3/watch.go [hunk 10]
```diff
@@ -65,6 +66,9 @@ type WatchResponse struct {
 	Created bool
 
 	closeErr error
+
+	// CancelReason is a reason of canceling watch
+	CancelReason string
```
Comment (heyitsanthony) [inline clientv3/watch.go]: `cancelReason`? data can be accessed through Err()

File: clientv3/watch.go [hunk 11]
```diff
@@ -85,6 +89,9 @@ func (wr *WatchResponse) Err() error {
 	case wr.CompactRevision != 0:
 		return v3rpc.ErrCompacted
 	case wr.Canceled:
+		if len(wr.CancelReason) != 0 {
+			return errors.New(wr.CancelReason)
```
Comment (heyitsanthony) [inline clientv3/watch.go]: should be able to reuse rpctypes errors here:
```go
return v3rpc.Error(grpc.Errorf(codes.FailedPrecondition, "%s", wr.cancelReason))
```

File: etcdserver/etcdserverpb/rpc.proto [hunk 12]
```diff
@@ -654,6 +654,9 @@ message WatchResponse {
   // watcher with the same start_revision again.
   int64 compact_revision  = 5;
 
+  // cancel_reason indicates a reason of cancel.
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: s/a reason of cancel/the reason for canceling the watcher./

File: etcdctl/ctlv3/command/watch_command.go [hunk 13]
```diff
@@ -129,6 +129,9 @@ func getWatchChan(c *clientv3.Client, args []string) (clientv3.WatchChan, error)
 
 func printWatchCh(ch clientv3.WatchChan) {
 	for resp := range ch {
+		if resp.Canceled {
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/watch_command.go]: `"watch was canceled (%v)\n"`?

File: etcdserver/api/v3rpc/watch.go [hunk 14]
```diff
@@ -180,6 +201,22 @@ func (sws *serverWatchStream) recvLoop() error {
...
+					Header:       sws.newResponseHeader(sws.watchStream.Rev()),
+					WatchId:      -1,
+					Canceled:     true,
+					CancelReason: rpctypes.ErrGRPCPermissionDenied.Error(),
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: add `Created: true` since it's a response to a watch create

File: clientv3/watch.go [hunk 15]
```diff
@@ -456,11 +463,23 @@ func (w *watchGrpcStream) run() {
 					wc.Send(ws.initReq.toPB())
 				}
 			case pbresp.Canceled:
-				delete(cancelSet, pbresp.WatchId)
-				if ws, ok := w.substreams[pbresp.WatchId]; ok {
-					// signal to stream goroutine to update closingc
-					close(ws.recvc)
-					closing[ws] = struct{}{}
+				if len(pbresp.CancelReason) != 0 {
```
Comment (heyitsanthony) [inline clientv3/watch.go]: client probably doesn't need this new path if the server sets `Created` to `true` when returning the error response

File: clientv3/watch.go [hunk 16]
```diff
@@ -605,7 +640,7 @@ func (w *watchGrpcStream) serveSubstream(ws *watcherStream, resumec chan struct{
 				return
 			}
 
-			if wr.Created {
+			if wr.Created || wr.Canceled {
```
Comment (heyitsanthony) [inline clientv3/watch.go]: `|| wr.Canceled` not needed if `Created` is `true` in the server response?

File: clientv3/watch.go (hunk 15 above)
Comment (mitake) [inline clientv3/watch.go]: I tried the idea of setting `Created` as `true`, but from the cancel reason couldn't propagated to a client. This is because `pbresp.WatchId` is -1 so the existing path of `pbresp.Created` cannot reply the watch response to the client. It is because handling the canceled response with the path of created will close `watchGrpcStream.revc` in `watchGrpcStream.addSubstream()`.
 This is why I added the new path and new function `dispatchCanceledEvent()`: https://github.com/coreos/etcd/pull/7809/files#diff-bdef56e462f289cae8e13ebbe16ba455R571

Of course the watching can be canceled but the client cannot know the reason. How do you think?

File: clientv3/watch.go (hunk 10 above)
Comment (mitake) [inline clientv3/watch.go]: Ah, it should be private. I'll fix it in the next update. Thanks.

File: clientv3/watch.go (hunk 11 above)
Comment (mitake) [inline clientv3/watch.go]: Thanks, I'll ueuse the code.

File: etcdserver/etcdserverpb/rpc.proto (hunk 12 above)
Comment (mitake) [inline etcdserver/etcdserverpb/rpc.proto]: I'll fix it in the next update.

File: etcdctl/ctlv3/command/watch_command.go (hunk 13 above)
Comment (mitake) [inline etcdctl/ctlv3/command/watch_command.go]: ditto

**Comment by mitake** [comment]
@heyitsanthony anyway, I updated minor coding style problems. The error propagation way are currently kept as before. If the cancel reason can be propagated with a more simpler way, of course I'll change it.

**Comment by heyitsanthony** [comment]
@mitake Seriously, it needs `Created: true` (see https://github.com/heyitsanthony/etcd/tree/auth-watch-created-true) or old clients will hang:

With created=true:
```sh
[anthony@etcd]$ ./bin/etcd-create-true &
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl user add root:root
User root created
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl auth enable
Authentication Enabled
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl-create-true watch abc
Error:  watch is canceled by the server
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl-3.1.8 watch abc
Error:  watch is canceled by the server
```
Without created=true:
```sh
[anthony@etcd]$ ./bin/etcd-no-create-true &
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl-3.1.8 watch abc
[hangs]
[anthony@etcd]$ ETCDCTL_API=3 ./bin/etcdctl-no-create-true -w simple watch abc
watch was canceled (rpc error: code = FailedPrecondition desc = rpc error: code = PermissionDenied desc = etcdserver: permission denied)
Error:  watch is canceled by the server
```

I'm more concerned about clients hanging than trying to propagate the specific error through the client at this point.

**Comment by mitake** [comment]
@heyitsanthony I see, the problem of hanging seems to be important. I picked and squashed your commit and added a todo of e2e (https://github.com/coreos/etcd/pull/7809/files#diff-12807139fa89b9ce1993cb698466b7ecR141). Could you take a look?

**Comment by heyitsanthony** [comment]
lgtm once CI greenlights. thanks!

**Comment by mitake** [comment]
@heyitsanthony thanks for your detailed review! CI passed so I'm merging it.
